File: frontend/de/unika/ipd/grgen/ast/exprevals/SourceExprNode.java
/*
 * GrGen: graph rewrite generator tool -- release GrGen.NET 3.6
 * Copyright (C) 2003-2013 Universitaet Karlsruhe, Institut fuer Programmstrukturen und Datenorganisation, LS Goos; and free programmers
 * licensed under LGPL v3 (see LICENSE.txt included in the packaging of this file)
 * www.grgen.net
 */

package de.unika.ipd.grgen.ast.exprevals;

import java.util.Collection;
import java.util.Vector;

import de.unika.ipd.grgen.ast.*;
import de.unika.ipd.grgen.ast.util.DeclarationResolver;
import de.unika.ipd.grgen.ast.util.DeclarationTypeResolver;
import de.unika.ipd.grgen.ir.Edge;
import de.unika.ipd.grgen.ir.IR;
import de.unika.ipd.grgen.ir.exprevals.SourceExpr;
import de.unika.ipd.grgen.parser.Coords;

/**
 * A node yielding the source node of an edge.
 */
public class SourceExprNode extends ExprNode {
	static {
		setName(SourceExprNode.class, "source expr");
	}

	private IdentNode edgeUnresolved;
	private IdentNode nodeTypeUnresolved;

	private EdgeDeclNode edgeDecl;
	private NodeTypeNode nodeType;
	
	public SourceExprNode(Coords coords, IdentNode edge, IdentNode nodeType) {
		super(coords);
		this.edgeUnresolved = edge;
		becomeParent(this.edgeUnresolved);
		this.nodeTypeUnresolved = nodeType;
		becomeParent(this.nodeTypeUnresolved);
	}

	/** returns children of this node */
	@Override
	public Collection<BaseNode> getChildren() {
		Vector<BaseNode> children = new Vector<BaseNode>();
		children.add(getValidVersion(edgeUnresolved, edgeDecl));
		children.add(getValidVersion(nodeTypeUnresolved, nodeType));
		return children;
	}

	/** returns names of the children, same order as in getChildren */
	@Override
	public Collection<String> getChildrenNames() {
		Vector<String> childrenNames = new Vector<String>();
		childrenNames.add("edge");
		childrenNames.add("nodeType");
		return childrenNames;
	}

	private static final DeclarationResolver<EdgeDeclNode> edgeResolver =
		new DeclarationResolver<EdgeDeclNode>(EdgeDeclNode.class);
	private static final DeclarationTypeResolver<NodeTypeNode> nodeTypeResolver =
		new DeclarationTypeResolver<NodeTypeNode>(NodeTypeNode.class);

	/** @see de.unika.ipd.grgen.ast.BaseNode#resolveLocal() */
	@Override
	protected boolean resolveLocal() {
		edgeDecl = edgeResolver.resolve(edgeUnresolved, this);
		nodeType = nodeTypeResolver.resolve(nodeTypeUnresolved, this);
		return edgeDecl!=null && nodeType!=null && getType().resolve();
	}

	/** @see de.unika.ipd.grgen.ast.BaseNode#checkLocal() */
	@Override
	protected boolean checkLocal() {
		return true;
	}

	@Override
	protected IR constructIR() {
		return new SourceExpr(edgeDecl.checkIR(Edge.class), getType().getType());
	}

	@Override
	public TypeNode getType() {
		return nodeType;
	}
	
	public boolean noDefElementInCondition() {
		if(edgeDecl.defEntityToBeYieldedTo) {
			edgeDecl.reportError("A def entity ("+edgeDecl+") can't be accessed from an if");
			return false;
		}
		return true;
	}
}

File: frontend/de/unika/ipd/grgen/ast/expr/graph/SourceExprNode.cs
/*
 * GrGen: graph rewrite generator tool -- release GrGen.NET 8.1
 * Copyright (C) 2003-2026 Universitaet Karlsruhe, Institut fuer Programmstrukturen und Datenorganisation, LS Goos; and free programmers
 * licensed under LGPL v3, some components/parts use different licenses (see LICENSE.txt included in the packaging of this file)
 * www.grgen.de / www.grgen.net
 */

namespace de.unika.ipd.grgen.ast.expr.graph
{

	using System.Collections.Generic;

	using de.unika.ipd.grgen.ast;
	using BuiltinFunctionInvocationBaseNode = de.unika.ipd.grgen.ast.expr.BuiltinFunctionInvocationBaseNode;
	using ExprNode = de.unika.ipd.grgen.ast.expr.ExprNode;
	using EdgeTypeNode = de.unika.ipd.grgen.ast.model.type.EdgeTypeNode;
	using NodeTypeNode = de.unika.ipd.grgen.ast.model.type.NodeTypeNode;
	using TypeNode = de.unika.ipd.grgen.ast.type.TypeNode;
	using de.unika.ipd.grgen.ast.util;
	using IR = de.unika.ipd.grgen.ir.IR;
	using Expression = de.unika.ipd.grgen.ir.expr.Expression;
	using SourceExpr = de.unika.ipd.grgen.ir.expr.graph.SourceExpr;
	using Coords = de.unika.ipd.grgen.parser.Coords;

	/// <summary>
	/// A node yielding the source node of an edge.
	/// </summary>
	public class SourceExprNode : BuiltinFunctionInvocationBaseNode
	{
		static SourceExprNode()
		{
			SetClassName(typeof(SourceExprNode), "source expr");
		}

		private ExprNode edge;

		private IdentNode nodeTypeUnresolved;
		private NodeTypeNode nodeType;

		public SourceExprNode(Coords coords, ExprNode edge, IdentNode nodeType)
			: base(coords)
		{
			this.edge = edge;
			BecomeParent(this.edge);
			this.nodeTypeUnresolved = nodeType;
			BecomeParent(this.nodeTypeUnresolved);
		}

		/// <summary>
		/// returns children of this node </summary>
		public override ICollection<BaseNode> Children
		{
			get
			{
				IList<BaseNode> children = new List<BaseNode>();
				children.Add(edge);
				children.Add(GetValidVersion(nodeTypeUnresolved, nodeType));
				return children;
			}
		}

		/// <summary>
		/// returns names of the children, same order as in getChildren </summary>
		public override ICollection<string> ChildrenNames
		{
			get
			{
				IList<string> childrenNames = new List<string>();
				childrenNames.Add("edge");
				childrenNames.Add("nodeType");
				return childrenNames;
			}
		}

		private static readonly DeclarationTypeResolver<NodeTypeNode> nodeTypeResolver =
				new DeclarationTypeResolver<NodeTypeNode>(typeof(NodeTypeNode));

		/// <seealso cref="de.unika.ipd.grgen.ast.BaseNode.resolveLocal() "/>
		protected internal override bool ResolveLocal()
		{
			nodeType = nodeTypeResolver.Resolve(nodeTypeUnresolved, this);
			return nodeType != null && Type.Resolve();
		}

		/// <seealso cref="de.unika.ipd.grgen.ast.BaseNode.checkLocal() "/>
		protected internal override bool CheckLocal()
		{
			if(!(edge.Type is EdgeTypeNode))
			{
				ReportError("The function source expects as argument (edgeToGetSourceNodeFrom) a value of type edge"
						+ " (but is given a value of type " + edge.Type.TypeName + ").");
				return false;
			}
			return true;
		}

		protected internal override IR ConstructIR()
		{
			edge = edge.Evaluate();
			return new SourceExpr(edge.CheckIR<Expression>(typeof(Expression)), Type.IRType);
		}

		public override TypeNode Type
		{
			get
			{
				return nodeType;
			}
		}
	}

}
